storage: TestReplicaDestroy highlights issues with marking replicas gc'ed

```
make stress PKG=./storage TESTS=TestReplicaDestroy

panic: replica 1 (range 1) was garbage collected

goroutine 23 [running]:
panic(0x549c5e0, 0xc820150820)
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/panic.go:481 +0x3e6
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/storage/store.go:2116 +0x3a5
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200d1b90, 0xc8201747f0)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
```

This error makes a lot of sense with our recent changes. We've made sure that nobody gets to use Raft any more, but now we're returning errors in code paths that don't expect errors at all (because they assume Raft "always just works"). Clearly we need an audit and more graceful handling in all those paths (at least for a sentinel-error-version of the above error).

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -238,7 +238,9 @@ type Replica struct {
 // initialized) Raft group. It assumes that the Replica lock is held.
 func (r *Replica) withRaftGroupLocked(f func(r *raft.RawNode) error) error {
 	if r.mu.destroyed != nil {
-		return r.mu.destroyed
+		// Silently ignore all operations on destroyed replicas. We can't return an
+		// error here as all errors returned from this method are considered fatal.
+		return nil
 	}
 
 	if r.mu.internalRaftGroup == nil {
@@ -1279,6 +1281,9 @@ func (r *Replica) proposeRaftCommand(
 	chan roachpb.ResponseWithError, func() bool, error) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
+	if r.mu.destroyed != nil {
+		return nil, nil, r.mu.destroyed
+	}
 	_, replica := r.mu.state.Desc.FindReplica(r.store.StoreID())
 	if replica == nil {
 		return nil, nil, roachpb.NewRangeNotFoundError(r.RangeID)

--- a/storage/replica_test.go
+++ b/storage/replica_test.go
@@ -4953,7 +4953,6 @@ func TestReplicaLoadSystemConfigSpanIntent(t *testing.T) {
 
 func TestReplicaDestroy(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	t.Skip("#7457")
 	tc := testContext{}
 	tc.Start(t)
 	defer tc.Stop()

--- a/storage/store_test.go
+++ b/storage/store_test.go
@@ -404,8 +404,15 @@ func TestStoreRemoveReplicaDestroy(t *testing.T) {
 	}
 
 	// Verify that removal of a replica marks it as destroyed so that future raft
-	// commands on the Replica structure will fail.
-	err = rng1.withRaftGroup(func(r *raft.RawNode) error { return nil })
+	// commands on the Replica will silently be dropped.
+	err = rng1.withRaftGroup(func(r *raft.RawNode) error {
+		return errors.Errorf("unexpectedly created a raft group")
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, _, err = rng1.proposeRaftCommand(context.TODO(), roachpb.BatchRequest{})
 	expected := "replica .* was garbage collected"
 	if !testutils.IsError(err, expected) {
 		t.Fatalf("expected error %s, but got %v", expected, err)

